### PR TITLE
Add broker log.

### LINKED ISSUE #188: Log broker activity in database
In order to be prepared for future changes in drivers, broker and image management, we should log all broker activities to know what happens.

Broker should log when:
- a machine is created
- a machine is terminated
- a session is opened
- a session is closed

These log needs to be stored in a new table which's schema needs to be defined.

The ultimate goal will be to generate a graph of broker's acitivty.

### PR DESCRIPTION
**Pull request by corentindrouet**
Fixes #188 
Broker add log when:
- A machine is created.
- A machine is terminated.
- A session is opened.
- A session is closed.
- A machine is assigned to a user.

### CODE REVIEW

**Comment by dynamiccast** [comment]
Please update documentation about brokerlogs

File: assets/app/router.js [hunk 1]
```diff
@@ -66,6 +66,7 @@ Router.map(function() {
       this.route('other-setting');
       this.route('look-and-feel');
     });
+    this.route('brokerlog', function() {});
```
Comment (dynamiccast) [inline assets/app/router.js:69]: I'm not sure the second parameter is needed.

**Comment by dynamiccast** [comment]
Tests only check for the presence of brokerlogs but do not check if data is actually correct. Please assert on the content with what you expect.

**Comment by dynamiccast** [comment]
It seems like there is often something link:

```
Brokerlog.create({
....
});
```

Maybe we should factorize this logging behavior in a `MachineService.log()` method.

File: assets/app/router.js (hunk 1 above)
Comment (corentindrouet) [reply assets/app/router.js:69]: without this parameter, front-end don't load the page.

Comment (dynamiccast) [reply assets/app/router.js:69]: @spacewoox what do you think ?

File: api/services/MachineService.js [hunk 2]
```diff
@@ -435,6 +460,28 @@ function getDefaultImage() {
...
 
+/*
+ * Create a new broker log
+ *
+ * @method createBrokerLog
+ * @param {Machine} the machine to log
+ * @param {string} the state to save (created, deleted, opened, ...)
+ * @return {Promise} created log
+ */
+function createBrokerLog(machine, state) {
```
Comment (dynamiccast) [inline api/services/MachineService.js]: If this function is not exported. It should be prefixed with an underscore.

File: api/services/MachineService.js [hunk 3]
```diff
@@ -241,14 +245,26 @@ function _createMachine() {
...
+          for (let i = 0; i < len; i++) {
+            if (_awaitingMachines[i] === machine) {
+              _awaitingMachines.splice(i, 1);
+            }
           }
-        }
-      });
+        })
+        .catch(() => {
+          _createBrokerLog({
```
Comment (dynamiccast) [inline api/services/MachineService.js]: `_createBrokerLog` should take an empty object as parameter if no machine is linked with the log record.

File: assets/app/protected/template.hbs [hunk 4]
```diff
@@ -112,6 +112,17 @@
               {{/link-to}}
             </li>
           {{/if}}
+
+          {{#if session.user.isAdmin}}
+            <li>
+              {{#link-to 'protected.brokerlog'}}
+                <span class="icon">
+                  <i class="material-icons">history</i>
```
Comment (dynamiccast) [inline assets/app/protected/template.hbs]: Use an other icon such as `th-list`.

File: tests/api/security.test.js [hunk 5]
```diff
@@ -2360,6 +2360,113 @@ module.exports = function() {
...
+
+        it('Admin should be authorized', function(done) {
+          nano.request(sails.hooks.http.app)
+            .get('/api/brokerlogs')
+            .set(nano.adminLogin())
+            .expect(200)
+            .end(done);
+        });
+
+        it('Regular users should be unauthorized', function(done) {
```
Comment (dynamiccast) [inline tests/api/security.test.js]: should be forbidden

File: tests/api/security.test.js [hunk 6]
```diff
@@ -2360,6 +2360,113 @@ module.exports = function() {
...
+          nano.request(sails.hooks.http.app)
+            .get('/api/brokerlogs')
+            .expect(401)
+            .end(done);
+        });
+      });
+
+      describe('Create a broker log', function () {
+
+        it('Admin should be authorized', function(done) {
```
Comment (dynamiccast) [inline tests/api/security.test.js]: Only internal logic should ever update information on this table. Only `find` and `findOne` should be accessible for admins.

**Comment by dynamiccast** [comment]
This PR needs more security tests for `findOne` and `update`.

**Comment by dynamiccast** [comment]
Brokerlog should be sorted by `createdAt` by default.

**Pull request merged**